.github/workflows/ci.yml

```
@@ -14,28 +14,57 @@ concurrency:
   cancel-in-progress: true
 jobs:
   build:
-    if: github.repository == 'pypi/warehouse'
-    runs-on: depot-ubuntu-24.04-arm
+    runs-on: ubuntu-24.04-arm
     outputs:
-      buildId: ${{ steps.build.outputs.build-id}}
+      buildId: ${{ github.run_id }}
     permissions:
-      id-token: write
+      packages: write
     steps:
       - name: Check out repository
         uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
         with:
           persist-credentials: false
-      - name: Set up Depot CLI
-        uses: depot/setup-action@b0b1ea4f69e92ebf5dea3f8713a1b0c37b2126a5 # v1.6.0
-      - name: Build image
-        id: build
-        uses: depot/build-push-action@636daae76684e38c301daa0c5eca1c095b24e780 # v1.14.0
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@b5ca514318bd6ebac0fb2aedd5d36ec1b5c232a2 # v3.10.0
+      - name: Cache
+        uses: actions/cache@5a3ec84eff668545956fd18022155c47e93e2684 # v4.2.3
+        id: cache
+        with:
+          path: |
+            var-cache-apt
+            var-lib-apt
+            root-cache-pip
+            root-npm
+          key: cache-${{ hashFiles('Dockerfile') }}
+      - name: inject cache into docker
+        uses: reproducible-containers/buildkit-cache-dance@653a570f730e3b9460adc576db523788ba59a0d7 # v3.2.0
+        with:
+          cache-map: |
+            {
+              "var-cache-apt": "/var/cache/apt",
+              "var-lib-apt": "/var/lib/apt",
+              "root-cache-pip": "/root/.cache/pip",
+              "root-npm": "/root/.npm"
+            }
+          skip-extraction: ${{ steps.cache.outputs.cache-hit }}
+      - name: Login To GHCR
+        uses: docker/login-action@74a5d142397b4f367a81961eba4e8cd7edddf772 # v3.4.0
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build and push
+        uses: docker/build-push-action@14487ce63c7a62a4a324b0bfb37086795e31c6c1 # v6.16.0
         with:
-          save: true
+          context: .
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
```

I think we need something like:
```suggestion
          cache-to: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.repo.full_name != github.repository && '' || 'type=gha,mode=max' }}
```
Otherwise action runs on PRs from untrusted forks can overwrite this cache as well.

(cc @woodruffw, should zizmor flag this?)

---

Huh yeah, it should -- I'm actually confused as to why it didn't, since I have a coordinate for the `push: true` part of this:

https://github.com/woodruffw/zizmor/blob/9790cec2727f04fee3cbaef1725f910c114e343e/src/audit/cache_poisoning.rs#L183-L186

---

Oh never mind, I see why it didn't: it considers `docker/build-push-action` a "publish sink" but not a "cache source" at the moment, since I'm only detecting `push: true` and not the `cache-to: ...` field.

This is a shortcoming in the current audit, I'll file an issue 🙂

---

https://github.com/woodruffw/zizmor/issues/774

---

Awesome, thanks! Just to confirm, this suggestion resolves the issue, correct?

---

I _think_ so, although I'm murky on whether doing the same for `cache-from` is also needed: that's a sink rather than a source so I _think_ it's okay, but I might be missing something.

---

I think we specifically want `cache-from` to always be permitted here, so PRs can take advantage of our central cache to speed up builds (but can't write to it).

---

Makes sense!